File: factory/java/AbstractFactory.java
import java.util.HashMap;
import java.util.Map;

interface IJavaReader {
    String read(String path);
}

interface IKotlinReader {
    String read(String path);
}

interface IScalaReader {
    String read(String path);
}

class XMLJavaReader implements IJavaReader {
    @Override
    public String read(String path) {
        return "(java) xml:" + path;
    }
}

class XMLKotlinReader implements IKotlinReader {
    @Override
    public String read(String path) {
        return "(kotlin) xml:" + path;
    }
}

class XMLScalaReader implements IScalaReader {
    @Override
    public String read(String path) {
        return "(scala) xml:" + path;
    }
}

class JSONJavaReader implements IJavaReader {
    @Override
    public String read(String path) {
        return "(java) json:" + path;
    }
}

class JSONKotlinReader implements IKotlinReader {
    @Override
    public String read(String path) {
        return "(kotlin) json:" + path;
    }
}

class JSONScalaReader implements IScalaReader {
    @Override
    public String read(String path) {
        return "(scala) json:" + path;
    }
}

class YAMLJavaReader implements IJavaReader {
    @Override
    public String read(String path) {
        return "(java) yaml:" + path;
    }
}

class YAMLKotlinReader implements IKotlinReader {
    @Override
    public String read(String path) {
        return "(kotlin) yaml:" + path;
    }
}

class YAMLScalaReader implements IScalaReader {
    @Override
    public String read(String path) {
        return "(scala) yaml:" + path;
    }
}

// ============================================================

interface IAbstractFactory {
    IJavaReader createJavaReader();

    IKotlinReader createKotlinReader();

    IScalaReader createScalaReader();
}

class XMLReaderAbstractFactory implements IAbstractFactory {
    @Override
    public IJavaReader createJavaReader() {
        return new XMLJavaReader();
    }

    @Override
    public IKotlinReader createKotlinReader() {
        return new XMLKotlinReader();
    }

    @Override
    public IScalaReader createScalaReader() {
        return new XMLScalaReader();
    }
}

class JSONReaderAbstractFactory implements IAbstractFactory {
    @Override
    public IJavaReader createJavaReader() {
        return new JSONJavaReader();
    }

    @Override
    public IKotlinReader createKotlinReader() {
        return new JSONKotlinReader();
    }

    @Override
    public IScalaReader createScalaReader() {
        return new JSONScalaReader();
    }
}

class YAMLReaderAbstractFactory implements IAbstractFactory {
    @Override
    public IJavaReader createJavaReader() {
        return new YAMLJavaReader();
    }

    @Override
    public IKotlinReader createKotlinReader() {
        return new YAMLKotlinReader();
    }

    @Override
    public IScalaReader createScalaReader() {
        return new YAMLScalaReader();
    }
}

// ============================================================

/**
 * 创建型 - 抽象工厂
 */
public class AbstractFactory {

    private static final Map<Integer, IAbstractFactory> cachedReaderAbstractFactories = new HashMap<>();

    static {
        // 工厂对象复用，缓存起来
        cachedReaderAbstractFactories.put(1, new XMLReaderAbstractFactory());
        cachedReaderAbstractFactories.put(2, new JSONReaderAbstractFactory());
        cachedReaderAbstractFactories.put(3, new YAMLReaderAbstractFactory());
    }

    static IAbstractFactory getReaderFactory(int type) {
        return cachedReaderAbstractFactories.get(type);
    }
}

File: factory/go/abstract_factory.go
package factory

import "fmt"

type IGoReader interface {
	Read(path string) string
}

type IRustReader interface {
	Read(path string) string
}

type XMLGoReader struct{}
type XMLRustReader struct{}
type JSONGoReader struct{}
type JSONRustReader struct{}
type YAMLGoReader struct{}
type YAMLRustReader struct{}

func (r *XMLGoReader) Read(path string) string {
	return "(go) xml:" + path
}

func (r *XMLRustReader) Read(path string) string {
	return "(rust) xml:" + path
}

func (r *JSONGoReader) Read(path string) string {
	return "(go) json:" + path
}

func (r *JSONRustReader) Read(path string) string {
	return "(rust) json:" + path
}

func (r *YAMLGoReader) Read(path string) string {
	return "(go) yaml:" + path
}

func (r *YAMLRustReader) Read(path string) string {
	return "(rust) yaml:" + path
}

// ============================================================

type IAbstractFactory interface {
	CreateGoReader() IGoReader
	CreateRustReader() IRustReader
}

type XMLReaderAbstractFactory struct{}
type JSONReaderAbstractFactory struct{}
type YAMLReaderAbstractFactory struct{}

func (f *XMLReaderAbstractFactory) CreateGoReader() IGoReader {
	return &XMLGoReader{}
}

func (f *XMLReaderAbstractFactory) CreateRustReader() IRustReader {
	return &XMLRustReader{}
}

func (f *JSONReaderAbstractFactory) CreateGoReader() IGoReader {
	return &JSONGoReader{}
}

func (f *JSONReaderAbstractFactory) CreateRustReader() IRustReader {
	return &JSONRustReader{}
}

func (f *YAMLReaderAbstractFactory) CreateGoReader() IGoReader {
	return &YAMLGoReader{}
}

func (f *YAMLReaderAbstractFactory) CreateRustReader() IRustReader {
	return &YAMLRustReader{}
}

// ============================================================

var cachedReaderAbstractFactories = map[int]IAbstractFactory{
	1: &XMLReaderAbstractFactory{},
	2: &JSONReaderAbstractFactory{},
	3: &YAMLReaderAbstractFactory{},
}

func GetReaderAbstractFactory(typ int) (IAbstractFactory, error) {
	if readerFactory, ok := cachedReaderAbstractFactories[typ]; ok {
		return readerFactory, nil
	}
	return nil, fmt.Errorf("unknown reader type: %d", typ)
}
